### ISSUE TITLE
Needs API Updates for unity 5

### ISSUE THREAD

**Issue by frankcarey** [googlesamples/cardboard-unity#17]
I"m getting a bunch of errors when trying to import into unity 5, probably API changes?:

Examples:

Assets/Cardboard/Editor/CardboardEditor.cs(40,9): error CS0117: `GUI' does not contain a definition for `changed'

Assets/Cardboard/Editor/CardboardEditor.cs(71,13): error CS0117: `GUI' does not contain a definition for `changed'

**Comment by smdol** [comment]
I was not able to reproduce this -- it imported fine and ran in Unity 5 for me.  Could you retry on an empty project?

**Comment by nosmirck** [comment]
I could import perfect, with no errors, the demo scene works fine in the editor, everything seems to work ok, but, when I compile and run in the device, it doesn't work... it's all black (some times white when moving the device) so I think the head tracking is working but not rendering correctly.

Is the plugin compatible with Unity5?

**Comment by smdol** [comment]
The all black screen is caused by the CardboardGUI script, which is somehow
not compatible with Unity 5.  Disable the script and the scene should work.

**Comment by nosmirck** [comment]
@smdol perfect! it works :)

**Comment by oatssss** [comment]
I created an empty scene in a new project and imported the cardboard package and agreed to go ahead with the automatic api updates Unity prompted for on import. I then went to disable the scripts within CardboardGUI to find that they were already disabled. So I built and ran the scene on my phone (HTC One M7) and it worked, however, if I looked in the opposite direction of the cube, as nosmirck mentioned, the screen would flash white, or display a strange white radial gradient. If you would like to see a screen recording of the issue, I'd be happy to make one.

I am also getting this warning:

>**Assets/Cardboard/Scripts/SkyboxMesh.cs(69,12): warning CS0618: UnityEngine.Renderer.castShadows' is obsolete: Property castShadows has been deprecated. Use shadowCastingMode instead.'**

If I choose not to go ahead with the automatic api updates that Unity prompts for on import of the package, these are the errors I get:

>**Assets/Cardboard/Scripts/CardboardEye.cs(68,23): error CS0619: UnityEngine.Component.camera' is obsolete: Property camera has been deprecated. Use GetComponent<Camera>() instead. (UnityUpgradable)'**

>**Assets/Cardboard/Scripts/CardboardEye.cs(68,30): error CS1061: Type UnityEngine.Component' does not contain a definition for rect' and no extension method rect' of type UnityEngine.Component' could be found (are you missing a using directive or an assembly reference?)**

When I try to import the cardboard package into an existing project and add CardboardMain to an existing scene, no matter what I do, the screen on my phone always stays black - even when CardboardGUI is non-existent in the scene. What I'm going to try next is rebuilding my scene in a new, empty project since it's quite a simple scene. I had no issues with this package prior to Unity 5.

Thank you for all your work thus far.

**Comment by jodybrewster** [comment]
+1

**Comment by smdol** [comment]
More details about the Unity 5 compatibility issues have been found recently.  The release notes here have been updated:
    https://developers.google.com/cardboard/unity/release-notes
See the "Known Issues" section under v0.4.5.

The latest info is that making any calls to GUI class methods (e.g. GUI.Button, etc) is what causes the black screen.  In the demo scene, removing CardboardGUI should disable all such calls, but in your own scene you might have some of these calls elsewhere in your code.

The build warnings and errors are from scripting changes in Unity 5.  The Cardboard scripts need to be updated to Unity 5 in order to fix those.  So far, I've been able to let the upgrade dialog do its thing without trouble.  

The SkyboxMesh (which gives the warning) is broken for a different reason:  Unity 5 skyboxes apparently no longer have 6 textures in them named _LeftTex, etc, so the script isn't finding them.  That has to be fixed too.

**Comment by smdol** [comment]
BTW, I've seen the white flash when looking where the GUI would normally be, but I've never seen the white radial gradient thing.  If you could capture that in a recording, post it here.

**Comment by oatssss** [comment]
Hey, good to hear you're working on it!! The flashing seems to just randomly put textures from the scene in front of the camera, I've also seen what looks to be part of the ground texture (the grid). Here's a video of it:

http://youtu.be/svERLby5gGs

Where the random flashing happens, the motion of the phone is just continuing the up and down pattern seen prior.

**Comment by darkesco** [comment]
Same issue here. I made a custom plane of a tile floor and added a material. The plane flashes in front of the camera at certain angles. I hope there is a fix soon :(

**Comment by VanKurt** [comment]
Me too. For me it looks even worse than what can be seen in the video posted above.
I'm also having several runtime exceptions (missing skybox textures).
A fix for Unity 5 (personal) would be great :-)

**Comment by smdol** [comment]
@darkesco @VanKurt 
Do your scenes have any GUI calls (GUI.Button, etc) or skybox?  If so, can you disable or remove them and test to see if that makes things render better?  More data will be helpful in tracking this down!

**Comment by VanKurt** [comment]
Nope. No GUI, no skybox.
Just some 3D models, a light and an instance of the CardboardMain prefab.

**Comment by frankcarey** [comment]
@VanKurt - (or others) maybe you can upload your projects to github if
they're simple? Then we might be able to debug it together. Unity requires
some configuration to be git-friendly. See
http://stackoverflow.com/questions/21573405/how-to-prepare-a-unity-project-for-git


Cheers,

Frank Carey

**Comment by VanKurt** [comment]
I have quickly put together minimalistic example that shows the error.
For simplicity I have uploaded it to Google Drive:
https://drive.google.com/file/d/0B7EyW7gfSuJGNzFwSDBHR1V6NXc/view?usp=sharing
(created on OS X)

**Comment by smdol** [comment]
Thanks @VanKurt for the test scene.  I think I have a workaround for the strange popping artifacts, although it also disables the barrel distortion and the UI stuff between the two images.

1) On the Cardboard.SDK object, disable the following options:
    Native Distortion Correction
    Alignment Marker
    Settings Button
2) Edit Cardboard.cs, in function EndOfFrame, comment out the line:
    GL.InvalidateState();

You may see some weird stuff around the edges of the phone and between the eyes.  That's because nothing in the SDK ever clears that area.  But these areas should not be visible when the phone is in a Cardboard.   The scene will look distorted in a Cardboard, just like it does in Unity 4 Free, because the distortion correction is turned off.  Hopefully this will still allow you to experiment while the real fix is being figured out.  

Near as I can tell, any time the Cardboard SDK calls functions on the GL class in Unity 5, things can go weird like this.  The above workaround just bypasses all the GL calls currently in the SDK.  Unity has been told, awaiting response.

Please confirm if this helps or not!

**Comment by mnbattaglia** [comment]
I was having the same black screen problem in my existing project, until I found I had an active script with a call to GUI.Label to draw current FPS. Removing that, skyboxes and CardboardGUI worked.
OTOH, I'm experiencing some flickering problems in my HUD (using Unity new UI), random camera jumps and other lighting problems.

Can't wait for the fix! Unity 5 seems really powerful and complete and having the distortion correction available in free version is great!!!

Thanks for this really useful SDK!

Martín (from Argentina)

**Comment by darkesco** [comment]
So I went back to Unity 4 and now it works great but there are about 6 sceens behind my 2 main screens. It actually works fine in the goggles, but it's not very professional looking outside the viewer. I imagine all those screens are chewing up the CPU/GPU as well. Anyway, I'm a noob, I'll screw with it till it works.

**Comment by smdol** [comment]
Note on skyboxes:  Unity 5 now has 3 skybox shaders:  procedural, cubed, and 6-sided (the Unity 4 kind).  The SkyboxMesh script only works with 6-sided skyboxes, as of v0.4.5.

**Comment by smdol** [comment]
Another skybox + Unity 5 update:  I'm seeing that Unity 5's skybox shaders use the active projection matrix now during rendering.  That means they should correctly handle the parallax caused by Cardboard's off-center projection, which was the whole reason for adding the SkyboxMesh script in the first place.  

I recommend ripping out SkyboxMesh script altogether, and trying the scene without it.  Make sure the sky seems very distant (farther than anything else in the scene) when viewed in stereo.

(I have to check lens flares to see if they are similarly treated.  If so, the StereoLensFlare is not needed anymore either.)

**Comment by smdol** [comment]
@darkesco  Add an extra Camera to the scene.  Set its properties like so:
clear flags = Solid Color
background = Black (or whatever)
culling mask = Nothing
depth = -100

See if that takes care of the glop around the two eye viewports.

**Comment by darkesco** [comment]
@smdol where do I place the camera in the hierarchy? I tried different places and nothing seemed to help. Do I need to associate it as the background renderer or anything... or is that with the depth -100 does? Reminder, I'm using Unity 4 free until 5 becomes usable with cardboard. Thanks for your help BTW

**Comment by darkesco** [comment]
Adding the extra camera worked! I had to add it from the created tab. The component menu didn't seem to work. It's all good now. Thanks smdol!

**Comment by smdol** [comment]
@oatssss Turning on the Development Build option in the Build dialog makes the flashing textures problem go away for me.  I've also removed all traces of GUI calls, and the SkyboxMesh script as well.  Can you see if it helps you?

**Comment by VanKurt** [comment]
Yeah, going into "Development Mode" solves all my issues, too. I wonder what this setting does? ;-)
Could I upload an app to the Play Store that was built using this setting? I guess I'll have to do some research...

**Comment by darkesco** [comment]
"Development mode is not available on games acquired through the App Store / Play Store" 

I'm sticking with Unity 4 for now. But it looks like Unity 4 free doesn't support reflections!?! I hope this is an oversight on my part.

**Comment by VanKurt** [comment]
Are there any official news from the Cardboard or Unity developers?
I'm really getting worried...

**Comment by darkesco** [comment]
@VanKurt I agree. I'm building anyway. Unity 4 looks like doo doo compared to Unity 5 IMO. I assume Google or Unity will get this resolved. Unreal Engine is great at VR but apk file size sucks and they don't appear to support cardboard.  Unity has the market it appears; let's see if they act.

**Comment by frankcarey** [comment]
I don't think these are fundamental issues, so don't freak out. I'm going
to try to get my own game working again today and will hopefully have a
pull request / fork that proves useful.


Cheers,

Frank Carey

**Comment by darkesco** [comment]
Thank you

**Comment by smdol** [comment]
@VanKurt Can you try an experiment?  Put in the extra Camera like I suggested to @darkesco upthread, and remove SkyboxMesh, StereoLensFlare, and all CardboardGUI stuff from the project.  Do not enable Development Build.  Does this fix your rendering issues?  (I've been noticing lately that I haven't had to enable this anymore.)

**Comment by VanKurt** [comment]
@smdol I tried adding a dummy camera as suggested by you. Also I have none of the scripts/components active (e.g. SkyboxMesh, LensFlare etc.). Still the rendering issues remain.
Only the Developer Build Mode fully solves the problems.

Another observation I made is that not all scenes are affected. I have one scene that works perfectly, and another that doesn't. Both use the same Cardboard camera system. Really strange...

**Comment by darkesco** [comment]
I'm still learning the subtleties of Unity. I'll try to remove GUIcalls. For the skybox, do I  just remove skybox.cs or do I need to remove the component on all 4 cameras? I have MainCamera, left  and right eye cameras, and the extra camera that clears the kaleidoscope issue.

**Comment by smdol** [comment]
@VanKurt That's what I'm seeing too, only now my scene not showing the problem was showing it before I made various changes to it.  That means the bug, wherever it is hiding, is triggered under very specific conditions, which aren't clear yet.  The experiment I had you try was a test for one possible condition (depth buffer not being properly cleared), but that doesn't seem to be it.

@darkesco The component to remove is SkyboxMesh, located on MainCamera alongside StereoController.

**Comment by bonkaa** [comment]
I foud some way to make the demo usable ith unity5 but it requires some patches.

A/ From the Hierarchy editor : 
	1/ select 'Main Camera' and set his tag to 'Untagged'
	2/ select 'CardboardMain' and set his tag to 'MainCamera'
	3/ select 'Cube' and open 'Teleport' script

B/ In 'Teleport.cs'
	1/replace :
	head = Camera.main.GetComponent<StereoController>().Head;
	with
	head = GameObject.Find("Head").GetComponent<CardboardHead>();
	
	2/ comment :
	//CardboardGUI.IsGUIVisible = true;
	//CardboardGUI.onGUICallback += this.OnGUI;

it removes the 'Reset' and 'Recenter' but the rest works fine with android

**Comment by smdol** [comment]
New version 0.4.9 is out.  It includes all the fixes we've discussed here:  Unity 5 import errors, the kaleidoscope around the outside edges, etc.  Note:  Development mode may still be needed, but double check.

**Comment by VanKurt** [comment]
Great news! Sadly the new version gives me a nasty error:

CommandInvokationFailure: Unable to convert classes into dex format. See the Console for details.
/Library/Java/JavaVirtualMachines/jdk1.7.0_60.jdk/Contents/Home/bin/java -Xmx2048M -Dcom.android.sdkmanager.toolsdir="/Users/borisbrock/Assets/AndroidSDK/tools" -Dfile.encoding=UTF8 -jar "/Applications/Unity/Unity.app/Contents/BuildTargetTools/AndroidPlayer/sdktools.jar" -

**Comment by smdol** [comment]
@VanKurt There are instructions for upgrading a project to the new version here:

https://developers.google.com/cardboard/unity/release-notes

**Comment by VanKurt** [comment]
@smdol That's exactly what I did: remove the old Cardboard folder, then import the new package. Did I miss anything else?

**Comment by smdol** [comment]
Yes, but I think the Release Notes are not clear.  Look down at the upgrade
instructions for v0.4.5 too.  There are some files to delete under Plugins.

**Comment by smdol** [comment]
@VanKurt The new release notes are changed to be clearer about upgrading.  Hope that helps!

**Comment by VanKurt** [comment]
@smdol Yes, it definitely will :-)
But now there's another problem: the CardboardSDKForUnity.unitypackage seems not to importable. Unity gives me a

Fatal Error
The file: 'MemoryStream' is corrupted. Remove it and launch Unity again!
[Position out of bounds!]

Before I was just copying the Cardboard-folder from the zip archive to my Unity project. So I did not see that error. I already tried several times, re-downloaded the package, restarted Unity etc.
BTW: I'm Using Unity 5.0 (main release) on OS X Yosemite.

**Comment by smdol** [comment]
I can honestly say I have never seen such an error!  If you have not
already tried, create a new project and import into that.  I will try it
too by downloading from Github myself.

**Comment by smdol** [comment]
Also, for anyone else reading: Don't just copy the Cardboard/ directory into your project.  It doesn't include the native libraries.  Those are in the Plugins/ directory which is only in the unitypackage.

**Comment by smdol** [comment]
@VanKurt Just noticed your update.  Good that it imports without error.  Does the Development Build option fix the flicker?

**Comment by VanKurt** [comment]
@smdol As with the previous version (0.4.5) of the Cardboard SDK, the flickering goes away when building in development mode.

**Comment by darkesco** [comment]
I saw an app released with the giant unity logo so I'm assuming that's unity 5. The full screen random texture flicker did not appear. I wonder how they got around that and were still able to release to the play store. Might try to gut the gui calls, but i have no idea to do that. I'll try searching for anything GUI and delete it. Will post back if I have any success.

**Comment by smdol** [comment]
It's a shot in the dark whether Development Build is needed, it seems.  Currently I can build all of my test projects in Unity 5, and the only one that needs Development Build is the original version of the DemoScene (the one now in the Legacy subdirectory).  The rest work fine without this option.  It sure is hard to figure out what the problem is in this kind of situation...

**Comment by VanKurt** [comment]
That's really strange. I would really love to publish my first (freeware!) VR-Game in the Play Store. So I keep praying to the flying spaghetti monster every night for a quick solution that works on EVERY Unity scene ;-)

**Comment by 999999333** [comment]
Hello, 
how to download previous version, during update i deleted it and now with a new version, unity 5 didnt work (blue screen)

**Comment by darkesco** [comment]
@smdol I've worked out my test scene build on the demo scene and am now ready to make a simple app for the play store. Iam planing to build the project from the compnents in the demo scene but not load the demo scene itself. I hope this will work. Btw: the texture that flickers is always the first texture of the first model, if that helps.

**Comment by darkesco** [comment]
I also appreciate all of your guys help. Im new and any help is awesome. You guy rock and so does VR!

**Comment by smdol** [comment]
@999999333 You select the tag corresponding to the previous release in the Branch: menu.  See attached picture.
![screen shot 2015-04-02 at 11 40 19 am](https://cloud.githubusercontent.com/assets/10277940/6970876/37ec711e-d92d-11e4-97a1-e74ece523f86.png)

**Comment by 999999333** [comment]
Thx, I found it. But in really for me it doesnt work in unity 5, so i reverted back to 0.4.5 and it works perfect (developers build checked).

**Comment by darkesco** [comment]
Issue is resolved with the new Unity update 5.0.1f1. I am still using 0.4.5 of the cardboard SDK. I have development mode unchecked and no texture flickering. Yay!

**Comment by VanKurt** [comment]
I just updated to 5.0.1f1 and the Cardboard 4.1.0 SDK.

But when running on the device, the app does not go into "VR mode", meaning no split screen, no sensor controlled looking etc. Just one simple camera facing´g to the front. In the Unity Editor everything seems fine...

**Comment by smdol** [comment]
Check the adb logcat for an exception being thrown or other error messages.

**Comment by VanKurt** [comment]
On startup I get the following messages. Not sure if they are bad/unusual though...
04-18 17:58:56.161    1603-1656/de.fun4mobile.brickbreakervr W/Adreno-EGL﹕ <qeglDrvAPI_eglGetConfigAttrib:607>: EGL_BAD_ATTRIBUTE
04-18 17:58:56.161    1603-1656/de.fun4mobile.brickbreakervr W/Adreno-EGL﹕ <qeglDrvAPI_eglGetConfigAttrib:607>: EGL_BAD_ATTRIBUTE
04-18 17:58:56.257    1603-1656/de.fun4mobile.brickbreakervr W/libc﹕ pthread_create sched_setscheduler call failed: Operation not permitted

**Comment by smdol** [comment]
No, those are normal, or at least not new.

**Comment by Pandaman3D** [comment]
Hey, just checking in to see if VanKurt's issue was ever fixed?  I'm fairly new to Unity, coming from a long line of Unreal...  But I too also can't go into VR Mode when I build onto my Android.  It works if I use the regular GoogleCardboardSDK.  But I can't even get the demo to transfer to my Android in VR mode.  I even created a new project with only importing your scripts.  Any fixes? :(  Thanks ahead of time!

**Comment by PHARTGAMES** [comment]
Hi guys,

I found this page while searching for a solution to the black screen on android problem.

Unity worked fine in vr mode previously, but tonight I started getting black screens (with the divider and the settings button) in every Unity cardboard app I have built..

I solved the problem by uninstalling the google cardboard app and deleting the associated data off the device.

Not sure why this worked, but it solved my issue.

I'm using a Note3 with 4.4.2 btw

Cheers.
